sips/sip-013/sip-013-semi-fungible-token-standard.md

```diff
@@ -0,0 +1,248 @@
+
+`(get-token-uri ((token-id uint)) (response (optional (string-ascii 256)) uint))`
+
+Returns an optional ASCII string that is a valid URI which resolves to this token type's metadata. These files can provide off-chain metadata about that particular token type, like descriptions, imagery, or any other information. The exact structure of the metadata is out of scope for this SIP. However, the metadata file should be in JSON format and should include a `sip` property containing a number:
+
+```JSON
+{
+	"sip": 16
+	// ... any other properties
+}
+```
+
+Applications consuming these metadata files can base display capabilities on the `sip` value. It should refer to a SIP number describing a JSON metadata standard.
+
+### Transfer
+
+`(transfer ((token-id uint) (amount uint) (sender principal) (recipient principal)) (response bool uint))`
+
+Transfer a token from the sender to the recipient. It is recommended to leverage Clarity primitives like `ft-transfer?` to help safeguard users. The function should return `(ok true)` on success or an `err` response containing an unsigned integer on failure. The failure codes follow the existing conventions of `stx-transfer?` and `ft-transfer?`.
+
+| Error code | Description                                      |
+|------------|--------------------------------------------------|
+| `u1`       | The sender has insufficient balance.             |
+| `u2`       | The sender and recipient are the same principal. |
+| `u3`       | Amount is `u0`.                                  |
+| `u4`       | The sender is not authorised to transfer tokens. |
+
+Error code `u4` is broad and may be returned under different cirumstances. For example, a token  contract with an allowance mechanism can return `(err u4)` when the `sender` parameter has no allowance for the specified token amount or if the sender is not equal to `tx-sender` or `contract-owner`. A token contract without an allowance mechanism can return `(err u4)` simply when the `sender` is not what is expected.
+
+Since it is possible for smart contracts to own tokens, it is recommended to check for both `tx-sender` and `contract-caller` as it allows smart contracts to transfer tokens it owns without having to resort to using `as-contract`. Such a guard can be constructed as follows:
+
+```clarity
+(asserts! (or (is-eq sender tx-sender) (is-eq sender contract-caller)) (err u4))
+```
+
+The `transfer` function should emit a special transfer event, as detailed in the Events section of this document.
+
+### Transfer with memo
+
+`(transfer-memo ((token-id uint) (amount uint) (sender principal) (recipient principal) (memo (buff 34))) (response bool uint))`
+
+Transfer a token from the sender to the recipient and emit a memo. This function follows the exact same procedure as `transfer` but emits the provided memo via `(print memo)`. The memo event should be the final event emitted by the contract. (See also the events section of this document below.)
+
+### Bulk transfers
```

Comment:
Thinking more about `transfer-many` and `transfer-many-memo`, I'm not 100% persuaded that they need to be part of the trait definition.  It might make more sense to publish a supplemental contract that takes a `sip-013` trait implementation and implements `transfer-many` and `transfer-many-memo` by iteratively calling the `transfer` function.  Then, each SIP 013 contract doesn't have to re-implement the same boilerplate each time (which exposes them to bugs).

Reply:
I would prefer that too if we had a `contract-call?` that does not reload the target contract into memory each time. Adding a `transfer-many` to the token contract is a lot more efficient at this point.

Reply:
I opened an issue for this here: https://github.com/stacks-network/stacks-blockchain/issues/3061

I agree that this is a problem, and it should ultimately be solved in the Stacks node by only charging the transaction once per loaded contract (or even charging the *block* once per loaded contract).  But, do you think the SIP could outline a migration plan so that `transfer-many` and `transfer-many-memo` could be deprecated in favor of this approach once Stacks 2.1 incorporates this fix?

Reply:
What if we define two separate traits?
One for all base SFT function, and another one for `transfer-many` and `transfer-many-memo`.

That way we'll be able to:
- support only base functions using `SIP-013-trait` 
- support base functions and transfer-many (all in one contract) using `SIP-013-trait` and `SIP-013-transfer-many-trait` 
- support base functions in token contract and transfer-many in different contract

Reply:
That's exactly what I was thinking too. We can do the same for SIP010.